### ISSUE TITLE
Policy roll-out in real-world results in the same action prediction at every timestep

### ISSUE THREAD

**Issue by milesprieb** [tonyzhaozh/act#6]
I am currently using the ACT model, similar to ALOHA, for learning a single-arm policy on the UR5e robot. I am seeing the same action prediction at every timestep, and it is an action that was never seen in any trajectories in the demonstrations. Besides obvious differences in the hardware, what model/training specific problems could I be facing based on this roll-out? Is it potentially related to the "style variable" (z)? I have provided my training results in screenshots below. Here are the hyperparameters used:

`python3 imitate_episodes.py \
--task_name rotate_bottle \
--ckpt_dir ./policy/ \
--policy_class ACT --kl_weight 10 --chunk_size 100 --hidden_dim 512 --batch_size 8 --dim_feedforward 3200 \
--num_epochs 5000 \
--lr 1e-5 \
--seed 0`

![Screenshot 2023-09-27 192155](https://github.com/tonyzhaozh/act/assets/61295777/625c8d18-9db0-427e-ab0e-a5700e39f013)

**Comment by tonyzhaozh** [comment]
Hi,

"I am seeing the same action prediction at every timestep"
Is it exactly the same action, or very similar actions? Can you double check if the policy can achieve the train/validation loss shown above with that prediction? e.g. you can plot its prediction on the train set. I would be quite surprised if predicting the same thing can get to 0.03 train and 0.13 val loss, given the data is normalized correctly with zero mean one std dev.

You can also consider increasing KL weight to 100, in case the latent is leaking information. Other things to try includes decreasing chunk_size, increasing batch size and increasing lr.

-Tony

**Comment by milesprieb** [comment]
I have resolved the issue. Turns out it actually was related to hardware-specific functionalities in my roll-out. Thank you for the speedy response and the great work! Hope to use it for some cool and novel tasks!

**Comment by Nimingez** [comment]
My situation is similar to yours. I collected 50 identical trajectories of UR. When inferring, there is a lag and it runs very slowly. I wonder if you have encountered a similar situation? @milesprieb

**Comment by tonyzhaozh** [comment]
Some suggestions:
- Try to time your inference latency [here](https://github.com/tonyzhaozh/act/blob/main/imitate_episodes.py#L249). I.e. add a time.time() before and after. The latency should be <0.05 with a good computer. In addition, try to time get_observation [here](https://github.com/tonyzhaozh/aloha/blob/main/aloha_scripts/real_env.py#L125) and it should be very small (<0.01s).
- If the policy inference latency is high, try disable --temporal_agg so policy inference happen every chunk_size steps. This will make the behavior more jerky but faster.

**Comment by anan-fighting** [comment]
> I have resolved the issue. Turns out it actually was related to hardware-specific functionalities in my roll-out. Thank you for the speedy response and the great work! Hope to use it for some cool and novel tasks!我已经解决了问题。结果发现，这其实和我推出时硬件相关的功能有关。感谢你的快速回复和出色的工作！希望能用它做一些有趣又新颖的任务！

Hello, I have recently been reproducing ACT on the UR5e single-arm platform. I have collected 31 pieces of data, but since the length of each data is different, I need to modify the part of the ACT source code related to data loading. Have you made any modifications?
The actual effect after reasoning is that the robotic arm keeps circling around the starting point and doesn't move down to grab the object. I don't know what the reason is.
